### DISCUSSION TITLE
There's a way to implement a token interceptor?

### DISCUSSION THREAD

**Discussion by veronesecoms** [TanStack/query#931] [General] [answered]
Hi, i need to create a token interceptor, i look in documentation and don't find a way to provide this.

I need to add a bearer token in all requests except login, someone knows if is possible to make this in react-query?

**Comment by jrozbicki** [comment, marked as answer]
This has really nothing to do with `react-query` itself. 

What you want to do is pass token to `Authorization` header. 

If you use `axios` you could do this in every request like:

```axios.get('http://someapi.com/api/todos', { headers: { Authorization: `Bearer ${token}`})```

but this get's repetetive pretty quickly.

Better solution would be to set up wrapper with global instance of axios and pass token there, this would look more like this:

```
// request.js
import axios from 'axios';
import { getToken } from 'someUtilsFunctions'

// optionaly add base url
const client = axios.create({ baseURL: 'http://someapi.com/api' });

const request = ({ ...options }) => {
    client.defaults.headers.common.Authorization = `Bearer ${getToken()}`;

    const onSuccess = (response) => response;
    const onError = (error) => {
       // optionaly catch errors and add some additional logging here
       return error;
    }

    return client(options).then(onSuccess).catch(onError);
}

export default request;
```

then you could wrap request (maybe besides /login if you don't want token there):

```
const getTodos = () => request({
   url: '/todos',
});
```

Having this you can use react-query as usual:

```
const { data } = useQuery('todos', async () => {
    const response = await getTodos();
    return response;
})
```

**Reply by inoumen** [reply]
To others who will use this code - pay attention that here is `baseUrl` instead of `baseURL`
I spend at least 30 minutes wondering why my base URL becomes 'localhost' that is definitely not what I declared :))

**Reply by itsmelion** [reply]
😂 that's why i like Typescript and linters.

**Reply by inoumen** [reply]
BTW, here the error is just returning, but you need to throw the error, otherwise, React Query will not understand that the error occurred.

**Reply by itsmelion** [reply]
True.
That's why its better to use the `onError()` method on the query.
Its better because i think it even awaits for retries etc. before failing.

**Reply by leslie555** [reply]
To others who will use this code in 2023，  pay attention that here is `baseURL` instead of `baseUrl`. 😂
![image](https://github.com/TanStack/query/assets/40735433/89c9a854-8162-4890-a94e-8c0c97f61013)

https://axios-http.com/docs/instance

**Reply by jrozbicki** [reply]
@leslie555 Thanks! I've updated original answer 🎉

**Comment by petarivanovv9** [comment]
As an addition to @jrozbicki 's answer, we could do something very similar. For instance, we could check the status of the error and navigate to the login page in our app. Here is the code:
```
import axios from 'axios';

import { API_URL } from '../config';

const SESSION_EXPIRED_STATUS_CODE = 401;

const baseApiClient = axios.create({
    baseURL: API_URL,
});

const apiClient = ({ ...options }) => {
    const onSuccess = (response) => response;

    const onError = (error) => {
        if (error.response.status === SESSION_EXPIRED_STATUS_CODE) {
            // Navigate to Login screen
        }

        return Promise.reject(error);
    };

    return baseApiClient(options)
        .then(onSuccess)
        .catch(onError);
};

export default apiClient;
```

In this way, we could use them like:
```
export const getUser = async () => {
    const { data } = await apiClient({
        url: `/user`,
    });
    return data;
};
```
and optionally we could add a hook:
```
export const useUser = (config) =>
    useQuery(['user'], getUser, config);
```

Hope that I've helped!

**Comment by itsmelion** [comment]
Anyhow, i think RQuery having an interceptor would be great.
especially having hooks on it, add mutation to consume refresh_tokens..
and better yet: get the token expire time, and set it as the staleTime of the user/auth query.
